src/System.Net.Http/tests/FunctionalTests/HttpProtocolTests.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.IO;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Xunit;
+
+namespace System.Net.Http.Functional.Tests
+{
+    public abstract class HttpProtocolTests
+    {
+        public virtual Stream GetStream(Stream s)
+        {
+            return s;
+        }
+
+        private bool GetHeaders(byte[] buffer, int length)
+        {
+            int offset = 0;
+            while (true)
+            {
+                int found = Array.IndexOf(buffer, (byte)'\r', offset, length - offset);
+                if (found < 0) return false;
+
+                if (found + 3 < length)
+                {
+                    if (buffer[found + 1] == (byte)'\n' &&
+                        buffer[found + 2] == (byte)'\r' &&
+                        buffer[found + 3] == (byte)'\n')
+                    {
+                        return true;
+                    }
+                }
+
+                offset = found + 1;
+            }
+        }
+        private Uri CreateServer(string response)
+        {
+            Socket listen = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+            listen.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+            listen.Listen(1);
+
+            listen.AcceptAsync().ContinueWith(async t =>
+            {
+                Socket accept = t.Result;
+                accept.NoDelay = true;
+                Stream s = GetStream(new NetworkStream(accept));
+
+                // Read whatever request the client sent us
+                // Note, we assume there's no content
+                byte[] buffer = new byte[4096];
+                int length = 0;
+                while (true)
+                {
+                    int bytesRead = await s.ReadAsync(buffer, length, buffer.Length - length);
+                    if (bytesRead == 0)
+                        throw new InvalidOperationException($"unexpected end of stream, buffer={Encoding.UTF8.GetString(buffer, 0, length)}");
+
+                    length += bytesRead;
+
+                    if (GetHeaders(buffer, length))
+                        break;
+                }
+
+                // Send the specified response
+                byte[] responseBytes = Encoding.UTF8.GetBytes(response);
+                await s.WriteAsync(responseBytes, 0, responseBytes.Length);
+                accept.Close();
+                listen.Close();
+            }, TaskContinuationOptions.ExecuteSynchronously);
```

Comment:
I think this is now the 3rd or 4th loopback server we have in these tests :)  Can we augment one of the existing ones however is needed rather than adding another?